flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/PythonKeyedCoProcessOperator.java

```diff
@@ -0,0 +1,314 @@
+            throws Exception {
+        long time = timer.getTimestamp();
+        Row timerKey = Row.of(timer.getKey());
+        if (procTime) {
+            reusableTimerData.setField(0, PROC_TIME_TIMER.value);
+        } else {
+            reusableTimerData.setField(0, EVENT_TIME_TIMER.value);
+        }
+        reusableTimerData.setField(1, time);
+        reusableTimerData.setField(2, timerService.currentWatermark());
+        reusableTimerData.setField(3, timerKey);
+        getRunnerInputTypeSerializer().serialize(reusableTimerData, baosWrapper);
+        pythonFunctionRunner.process(baos.toByteArray());
+        baos.reset();
+        elementCount++;
+        checkInvokeFinishBundleByCount();
+        emitResults();
+    }
+
+    /**
+     * Handler the timer registration request from python user defined function. Before registering
+     * the timer, we must set the current key to be the key when the timer is register in python
+     * side.
+     *
+     * @param runnerOutput The timer registration request data.
+     */
+    private void registerTimer(Row runnerOutput) {
+        synchronized (getKeyedStateBackend()) {
+            byte type = (byte) runnerOutput.getField(0);
+            long time = (long) runnerOutput.getField(1);
+            Object timerKey = ((Row) (runnerOutput.getField(2))).getField(0);
+            setCurrentKey(timerKey);
+            if (type == REGISTER_EVENT_TIMER.value) {
+                this.timerService.registerEventTimeTimer(time);
+            } else if (type == REGISTER_PROC_TIMER.value) {
+                this.timerService.registerProcessingTimeTimer(time);
+            } else if (type == DEL_EVENT_TIMER.value) {
+                this.timerService.deleteEventTimeTimer(time);
+            } else if (type == DEL_PROC_TIMER.value) {
+                this.timerService.deleteProcessingTimeTimer(time);
+            }
+        }
+    }
+
+    private static TypeInformation<Row> constructKeyTypeInfo(TypeInformation<Row> inputTypeInfo) {
+        return new RowTypeInfo(((RowTypeInfo) inputTypeInfo).getTypeAt(0));
+    }
+
+    private static TypeInformation<Row> constructRunnerInputTypeInfo(
+            TypeInformation<Row> inputTypeInfo1,
+            TypeInformation<Row> inputTypeInfo2,
+            TypeInformation<Row> keyTypeInfo) {
+        // structure: [isLeftUserInput, leftInput, rightInput]
+        RowTypeInfo unifiedInputTypeInfo =
+                new RowTypeInfo(Types.BOOLEAN, inputTypeInfo1, inputTypeInfo2);
+
+        // structure: [isTimerTrigger, timestamp, currentWatermark, key, userInput]
+        return Types.ROW(Types.BYTE, Types.LONG, Types.LONG, keyTypeInfo, unifiedInputTypeInfo);
+    }
+
+    private static TypeInformation<Row> constructRunnerOutputTypeInfo(
+            TypeInformation<?> outputTypeInfo, TypeInformation<Row> keyTypeInfo) {
+        // structure: [isTimerRegistration, timestamp, key, userOutput]
```

Comment:
The fist field has four values `register event timer`, `register pro timer`, `del event timer` and `del pro timer`, so using `isTimerRegistration` is not accurate